programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,303 @@
+    ) -> Result<()> {
+        require!(!self.state.paused, MarinadeError::ProgramIsPaused);
+        require!(
+            !self.state.withdraw_stake_account_enabled,
+            MarinadeError::WithdrawStakeAccountIsNotEnabled
+        );
+
+        // record  for event
+        let user_msol_balance = self.burn_msol_from.amount;
+        // save msol price source
+        let total_virtual_staked_lamports = self.state.total_virtual_staked_lamports();
+        let msol_supply = self.state.msol_supply;
+
+        check_token_source_account(
+            &self.burn_msol_from,
+            self.burn_msol_authority.key,
+            msol_amount,
+        )
+        .map_err(|e| e.with_account_name("burn_msol_from"))?;
+
+        let mut stake = self.state.stake_system.get_checked(
+            &self.stake_list.data.as_ref().borrow(),
+            stake_index,
+            self.stake_account.to_account_info().key,
+        )?;
+        let last_update_stake_delegation = stake.last_update_delegated_lamports;
+
+        // require the stake is not in emergency_unstake
+        require_neq!(
+            stake.is_emergency_unstaking,
+            0,
+            MarinadeError::StakeAccountIsEmergencyUnstaking
+        );
+
+        // require stake is active (deactivation_epoch == u64::MAX)
+        let delegation = self.stake_account.delegation().ok_or_else(|| {
+            error!(MarinadeError::RequiredDelegatedStake).with_account_name("stake_account")
+        })?;
+        require_eq!(
+            delegation.deactivation_epoch,
+            std::u64::MAX,
+            MarinadeError::RequiredActiveStake
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check currently_staked in this account & validator vote-key
+        check_stake_amount_and_validator(
+            &self.stake_account,
+            stake.last_update_delegated_lamports,
+            &validator.validator_account,
+        )?;
+
+        // compute how many lamport the burned mSOL represents
+        let sol_value = self.state.msol_to_sol(msol_amount)?;
+        require_gt!(sol_value, self.state.min_withdraw);
+
+        // apply withdraw_stake_account_fee to avoid economical attacks
+        // withdraw_stake_account_fee must be >= one epoch staking rewards
+        let withdraw_stake_account_fee_lamports =
+            self.state.withdraw_stake_account_fee.apply(sol_value);
+        // the fee value will be burned but not delivered, thus increasing mSOL value slightly for all mSOL holders
+        let split_lamports = sol_value - withdraw_stake_account_fee_lamports;
+        // check withdraw amount (stake) >= self.state.stake_system.min_stake
+        // and also >= self.state.min_withdraw
+        require_gte!(
+            split_lamports,
+            std::cmp::max(self.state.min_withdraw, self.state.stake_system.min_stake),
```

Comment:
Don't need to check against `self.state.min_withdraw` twice. Please check only for min_stake here and use another error code